File: spring-webflux/src/main/java/victor/training/reactive/usecase/grouping/Apis.java
package victor.training.reactive.usecase.grouping;

import lombok.extern.slf4j.Slf4j;
import reactor.core.publisher.Mono;

import java.time.Duration;
import java.util.List;

@Slf4j
public class Apis {

   public Mono<Void> apiA(Integer oddItem) {
      return api("A", oddItem);
   }
   public Mono<Void> apiB(Integer oddItem) {
      return api("B", oddItem);
   }

   public Mono<Void> apiC(List<Integer> evenItemPage) {
      return api("C", evenItemPage);
   }

   private Mono<Void> api(String apiName, Object data) {
      return Mono.delay(Duration.ofMillis(100))
          .doOnSubscribe(s -> log.info("Api" + apiName + ": {} START", data))
          .doOnNext(item -> log.info("Api" + apiName + ": {} END", data))
          .then();
   }
}

File: spring-webflux/src/main/java/victor/training/reactive/usecase/grouping/Apis.kt
package victor.training.reactive.usecase.grouping

import org.reactivestreams.Subscription
import org.slf4j.LoggerFactory
import reactor.core.publisher.Mono
import java.time.Duration

open class Apis {
    fun apiA(oddItem: Int): Mono<Void> {
        return api("A", oddItem)
    }

    fun apiB(oddItem: Int): Mono<Void> {
        return api("B", oddItem)
    }

    fun apiC(evenItemPage: List<Int>): Mono<Void> {
        return api("C", evenItemPage)
    }

    private fun api(apiName: String, data: Any): Mono<Void> {
        return Mono.delay(Duration.ofMillis(100))
            .doOnSubscribe { s: Subscription? ->
                log.info(
                    "Api$apiName: {} START", data
                )
            }
            .doOnTerminate { log.info("Api$apiName: {} END", data) }
            .then()
    }

    companion object {
        private val log = LoggerFactory.getLogger(Apis::class.java)
    }
}